File: toolpacks/gamecube-decomp/research/ghidra/runners/ghidra_scripts/ExportXrefs.java
// Export every reference in the current program as JSON Lines.
// @category GameCube Decomp

import java.io.BufferedWriter;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.symbol.RefType;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceIterator;
import ghidra.program.model.symbol.ReferenceManager;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolTable;

public class ExportXrefs extends GhidraScript {

	private static final char[] HEX = "0123456789abcdef".toCharArray();

	@Override
	protected void run() throws Exception {
		String[] scriptArgs = getScriptArgs();
		if (scriptArgs.length < 1) {
			throw new IllegalArgumentException("ExportXrefs.java requires an output path argument");
		}

		Path outputPath = Paths.get(scriptArgs[0]);
		ReferenceManager referenceManager = currentProgram.getReferenceManager();
		FunctionManager functionManager = currentProgram.getFunctionManager();
		SymbolTable symbolTable = currentProgram.getSymbolTable();
		ReferenceIterator references =
			referenceManager.getReferenceIterator(currentProgram.getMinAddress());
		int count = 0;

		try (BufferedWriter output =
			Files.newBufferedWriter(outputPath, StandardCharsets.UTF_8)) {
			while (references.hasNext()) {
				Reference reference = references.next();
				Address from = reference.getFromAddress();
				Address to = reference.getToAddress();
				String fromAddress = addressText(from);
				String toAddress = addressText(to);
				RefType referenceType = reference.getReferenceType();
				String referenceTypeText = referenceType.toString();
				String fromSymbol = fromSymbolName(functionManager, symbolTable, from);
				String toSymbol = toSymbolName(functionManager, symbolTable, to);
				String text = valueOrEmpty(fromSymbol) + " " + referenceTypeText + " " +
					valueOrEmpty(toSymbol) + " " + fromAddress + " " + toAddress;

				output.write(jsonRow(fromAddress, toAddress, referenceTypeText,
					referenceType.isCall(), referenceType.isData(), fromSymbol, toSymbol, text));
				output.newLine();
				count++;
			}
		}

		println("EXPORT_XREFS_SUMMARY count=" + count + " output=" + outputPath);
	}

	private static String addressText(Address address) {
		return "0x" + address;
	}

	private static String primarySymbolName(SymbolTable symbolTable, Address address) {
		Symbol symbol = symbolTable.getPrimarySymbol(address);
		return symbol == null ? null : symbol.getName();
	}

	private static String fromSymbolName(FunctionManager functionManager,
			SymbolTable symbolTable, Address address) {
		Function function = functionManager.getFunctionContaining(address);
		return function == null ? primarySymbolName(symbolTable, address) : function.getName();
	}

	private static String toSymbolName(FunctionManager functionManager,
			SymbolTable symbolTable, Address address) {
		Function function = functionManager.getFunctionAt(address);
		return function == null ? primarySymbolName(symbolTable, address) : function.getName();
	}

	private static String valueOrEmpty(String value) {
		return value == null ? "" : value;
	}

	private static String jsonRow(String fromAddress, String toAddress, String referenceType,
			boolean isCall, boolean isData, String fromSymbol, String toSymbol, String text) {
		StringBuilder row = new StringBuilder(256);
		row.append("{\"from_address\": ").append(jsonString(fromAddress));
		row.append(", \"from_symbol\": ").append(jsonString(fromSymbol));
		row.append(", \"id\": ").append(jsonString("xref:" + fromAddress + ":" + toAddress));
		row.append(", \"is_call\": ").append(isCall);
		row.append(", \"is_data\": ").append(isData);
		row.append(", \"kind\": \"ghidra_xref\"");
		row.append(", \"ref_type\": ").append(jsonString(referenceType));
		row.append(", \"text\": ").append(jsonString(text));
		row.append(", \"to_address\": ").append(jsonString(toAddress));
		row.append(", \"to_symbol\": ").append(jsonString(toSymbol));
		row.append('}');
		return row.toString();
	}

	private static String jsonString(String value) {
		if (value == null) {
			return "null";
		}

		StringBuilder escaped = new StringBuilder(value.length() + 2);
		escaped.append('\"');
		for (int index = 0; index < value.length(); index++) {
			char character = value.charAt(index);
			switch (character) {
				case '\"':
					escaped.append("\\\"");
					break;
				case '\\':
					escaped.append("\\\\");
					break;
				case '\b':
					escaped.append("\\b");
					break;
				case '\f':
					escaped.append("\\f");
					break;
				case '\n':
					escaped.append("\\n");
					break;
				case '\r':
					escaped.append("\\r");
					break;
				case '\t':
					escaped.append("\\t");
					break;
				default:
					if (character < 0x20 || character > 0x7e) {
						escaped.append("\\u");
						escaped.append(HEX[(character >>> 12) & 0xf]);
						escaped.append(HEX[(character >>> 8) & 0xf]);
						escaped.append(HEX[(character >>> 4) & 0xf]);
						escaped.append(HEX[character & 0xf]);
					}
					else {
						escaped.append(character);
					}
			}
		}
		escaped.append('\"');
		return escaped.toString();
	}
}

File: toolpacks/gamecube-decomp/research/ghidra/runners/ghidra_scripts/ExportXrefs.py
# Export every reference in the current program as JSON Lines.
# @category GameCube Decomp

import json


def address_text(address):
    return "0x" + str(address)


def primary_symbol_name(symbol_table, address):
    symbol = symbol_table.getPrimarySymbol(address)
    if symbol is None:
        return None
    return str(symbol.getName())


def from_symbol_name(function_manager, symbol_table, address):
    function = function_manager.getFunctionContaining(address)
    if function is not None:
        return str(function.getName())
    return primary_symbol_name(symbol_table, address)


def to_symbol_name(function_manager, symbol_table, address):
    function = function_manager.getFunctionAt(address)
    if function is not None:
        return str(function.getName())
    return primary_symbol_name(symbol_table, address)


script_args = getScriptArgs()
if len(script_args) < 1:
    raise ValueError("ExportXrefs.py requires an output path argument")

output_path = script_args[0]
reference_manager = currentProgram.getReferenceManager()
function_manager = currentProgram.getFunctionManager()
symbol_table = currentProgram.getSymbolTable()
references = reference_manager.getReferenceIterator(currentProgram.getMinAddress())
count = 0
output_file = open(output_path, "w")

try:
    while references.hasNext():
        reference = references.next()
        from_address = address_text(reference.getFromAddress())
        to_address = address_text(reference.getToAddress())
        ref_type = reference.getReferenceType()
        ref_type_text = str(ref_type)
        from_symbol = from_symbol_name(
            function_manager, symbol_table, reference.getFromAddress()
        )
        to_symbol = to_symbol_name(
            function_manager, symbol_table, reference.getToAddress()
        )
        row = {
            "id": "xref:%s:%s" % (from_address, to_address),
            "kind": "ghidra_xref",
            "from_address": from_address,
            "to_address": to_address,
            "ref_type": ref_type_text,
            "is_call": bool(ref_type.isCall()),
            "is_data": bool(ref_type.isData()),
            "from_symbol": from_symbol,
            "to_symbol": to_symbol,
            "text": "%s %s %s %s %s"
            % (from_symbol or "", ref_type_text, to_symbol or "", from_address, to_address),
        }
        output_file.write(json.dumps(row, sort_keys=True))
        output_file.write("\n")
        count += 1
finally:
    output_file.close()

print "EXPORT_XREFS_SUMMARY count=%d output=%s" % (count, output_path)
